### PR TITLE
[SPARK-42555][CONNECT] Add JDBC to DataFrameReader

### PR DESCRIPTION
**Pull request by beliefer**
### What changes were proposed in this pull request?
Currently, the connect project have the new `DataFrameReader` API which is corresponding to Spark `DataFrameReader` API. But the connect `DataFrameReader` missing the jdbc API.

### Why are the changes needed?
This PR try to add JDBC to `DataFrameReader`

### Does this PR introduce _any_ user-facing change?
'No'.
New feature.

### How was this patch tested?
New test cases.

### CODE REVIEW

**Comment by beliefer** [comment]
The jdbc API seems hard to test, do we need a test case? @hvanhovell @HyukjinKwon @zhengruifeng @dongjoon-hyun

**Comment by beliefer** [comment]
There is another kind jdbc API, see: https://github.com/apache/spark/blob/79da1ab400f25dbceec45e107e5366d084138fa8/sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala#L316
I will create another PR to add proto msg and implement it.

**Comment by zhengruifeng** [comment]
I guess you can refer to `JDBCSuite` and `ClientE2ETestSuite` ?

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/DataFrameReader.scala [hunk 1]
```diff
@@ -184,6 +186,67 @@ class DataFrameReader private[sql] (sparkSession: SparkSession) extends Logging
...
+   * @param table Name of the table in the external database.
+   * @param columnName Alias of `partitionColumn` option. Refer to `partitionColumn` in
+   *                   <a href="https://spark.apache.org/docs/latest/sql-data-sources-jdbc.html#data-source-option">
+   *                     Data Source Option</a> in the version you use.
+   * @param connectionProperties JDBC database connection arguments, a list of arbitrary string
+   *                             tag/value. Normally at least a "user" and "password" property
+   *                             should be included. "fetchsize" can be used to control the
+   *                             number of rows per fetch and "queryTimeout" can be used to wait
+   *                             for a Statement object to execute to the given number of seconds.
+   * @since 3.5.0
```
Comment (LuciferYang) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/DataFrameReader.scala:230]: Will this not be backport to 3.4.0?

Comment (hvanhovell) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/DataFrameReader.scala:230]: I am fine with marking this as 3.4.0. There is no harm in adding this to 3.4.0, if the RC does not pass.

**Comment by hvanhovell** [comment]
@beliefer you can create a test in `PlanGenerationTestSuite`. That will at least validate the proto message we are generating, and it will validate that plan you are producing yields a valid plan in `ProtoToPlanTestSuite`.

**Comment by beliefer** [comment]
> I guess you can refer to `JDBCSuite` and `ClientE2ETestSuite` ?

The built-in H2 running in server side and we can't start H2 database at connect API.

**Comment by beliefer** [comment]
> @beliefer you can create a test in `PlanGenerationTestSuite`. That will at least validate the proto message we are generating, and it will validate that plan you are producing yields a valid plan in `ProtoToPlanTestSuite`.

OK. I added the test cases. But throws
```
org.h2.jdbc.JdbcSQLSyntaxErrorException: Schema "TEST" not found; SQL statement:
SELECT * FROM TEST.TIMETYPES WHERE 1=0 [90079-214]
```
 when ProtoToParsedPlanTestSuite validating the golden file. The built-in H2 running in server side and we can't start H2 database at connect API.
@hvanhovell Could you tell me how to fix or avoid it ?

File: connector/connect/server/src/test/scala/org/apache/spark/sql/connect/ProtoToParsedPlanTestSuite.scala [hunk 2]
```diff
@@ -57,6 +59,30 @@ import org.apache.spark.sql.util.CaseInsensitiveStringMap
  */
 // scalastyle:on
 class ProtoToParsedPlanTestSuite extends SparkFunSuite with SharedSparkSession {
+  val url = "jdbc:h2:mem:testdb0"
+  val urlWithUserAndPass = "jdbc:h2:mem:testdb0;user=testUser;password=testPass"
+  var conn: java.sql.Connection = null
```
Comment (hvanhovell) [inline connector/connect/server/src/test/scala/org/apache/spark/sql/connect/ProtoToParsedPlanTestSuite.scala:64]: Do we need to close the connection?

**Review by hvanhovell — COMMENTED** [review]
Almost good to go! Left a few comments.

File: connector/connect/server/src/test/scala/org/apache/spark/sql/connect/ProtoToParsedPlanTestSuite.scala (hunk 2 above)
Comment (beliefer) [reply connector/connect/server/src/test/scala/org/apache/spark/sql/connect/ProtoToParsedPlanTestSuite.scala:64]: Yeah. I forgot it.

**Review by hvanhovell — APPROVED** [review]
LGTM

**Review by dongjoon-hyun — COMMENTED** [review]
Hi, @beliefer .
- Could you build and test Scala 2.13 manually?
- Could you apply scalafmt? It seems that the CI complains about that.

**Comment by beliefer** [comment]
@dongjoon-hyun @hvanhovell It seems the build scala 2.13 failed is unrelated to this PR.

**Comment by hvanhovell** [comment]
merging to master/3.4

**Comment by beliefer** [comment]
@hvanhovell @dongjoon-hyun @LuciferYang Thank you!

**Pull request closed without merging**